text/0113-virtual-script-container.md

```diff
@@ -0,0 +1,207 @@
+
+To ensure we can solve the problem while continuing to guarantee the benefits that are provided by the synthetic-shadow polyfill for all other components running on the main window, we are currently proposing the following characteristics for the virtual-script design:
+
+* Create a JS sandbox before the Synthetic Shadow Polyfill is evaluated.
+  * The sandbox is created using a same-domain iframe.
+
+* Use this JS sandbox to virtualize the DOM APIs in order to prevent these APIs to be subject to the Synthetic Shadow Polyfill patches:
+  * This can be done by caching references to the to-be-patched APIs inside the sandbox before the synthetic shadow is evaluated.
+
+* Evaluate one or more third party libraries inside this sandbox
+  * This allows the code to run in a non-patched environment, while the rest of the document is running synthetic shadow.
+
+* Package the above functionality in a web component which semantically matches the `<script>` tag API.
+
+* Provide an additional mechanism to remap globals when needed.
+    * The sandbox does NOT include any global name defined in the main window, whether it is defined before or after the creation of the sandbox. But the `<virtual-script>` can opt-in to share a global value between the sandbox and the main window, this must be done via the `extra-globals` argument passed to the `<virtual-script extra-globals="dataLayer">` element.
+
+Other considerations for Phase I:
+
+ * due to the explicit order of evaluation required for this to function, we plan to only support one sandbox that can be shared by multiple virtual scripts.
+ * budget is around 5kb (minify/gzip).
+ * performance penalties will be added to the virtual scripts as they run inside a membrane:
+   * evaluation budget: 30% slower than evaluating the same script in the main window.
+   * operation budget: 50% slower than executing the same task is the main window.
+   * 3rd party scripts are rarely part of the critical path for the application to function.
+
+### Phase II: Native Shadow
+
+For applications which are using the native shadow, the underlying implementation will be different, a more complicated one. We are currently proposing the following characteristics for the virtual-script design:
+
+* During the evaluation of the virtual script component class, patch `Element.prototype.attachShadow`, to record any shadow created in the app.
```

Comment:
Patching `Element.prototype.attachShadow` will not be sufficient to intercept all the shadow trees creations if the application is using the declarative shadow DOM syntax.

We can wait for the `DOMContentLoaded` event and retrieve the list of all the open shadow trees created via the declarative shadow DOM syntax as part of the initial page load. I don't there will be an easy way to find all the other shadow trees created via the declarative shadow after that. But I guess it's fine.

Reply:
I was thinking about this as well... there are other issues. lets chat about it.